### PR TITLE
Introduce a "task" selection option to AdminUpdate for running just the ARO Operator update steps

### PR DESCRIPTION
**Pull request by hawkowl**
### Which issue this PR addresses:

Part of https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/10430645/?workitem=10609866

### What this PR does / why we need it:

This allows the ARO Operator to update without running the rest of the AdminUpdate. The Operator should always be updated, even if some parts of the Azure infrastructure aren't/can't be (e.g. DNS) since the Operator will be a mechanism to fix or notify the customer of some of these issues.

### Test plan for issue:

Tests in PR

### Is there any documentation that needs to be updated for this PR?

maybe?

### CODE REVIEW

File: pkg/backend/openshiftcluster.go [hunk 1]
```diff
@@ -266,14 +275,17 @@ func (ocb *openShiftClusterBackend) endLease(ctx context.Context, log *logrus.En
...
-	if doc.OpenShiftCluster.Properties.ProvisioningState == api.ProvisioningStateAdminUpdating {
+	if doc.OpenShiftCluster.Properties.ProvisioningState.IsAdminOperation() {
 		provisioningState = doc.OpenShiftCluster.Properties.LastProvisioningState
 		failedProvisioningState = doc.OpenShiftCluster.Properties.FailedProvisioningState
 
-		if backendErr == nil {
-			adminUpdateError = to.StringPtr("")
-		} else {
-			adminUpdateError = to.StringPtr(backendErr.Error())
+		// Don't update adminUpdateError on OperatorUpdate
```
Comment (m1kola) [inline pkg/backend/openshiftcluster.go:282]: Why not?

File: pkg/api/v20191231preview/openshiftcluster.go [hunk 2]
```diff
@@ -79,6 +81,13 @@ const (
 	ProvisioningStateFailed        ProvisioningState = "Failed"
 )
 
+func ToExternalProvisioningState(state api.ProvisioningState) ProvisioningState {
+	if state == "OperatorUpdating" {
+		return ProvisioningStateAdminUpdating
+	}
+	return ProvisioningState(state)
+}
```
Comment (m1kola) [inline pkg/api/v20191231preview/openshiftcluster.go:89]: I somehow feel that this is a bad idea to have 1 external provisioning state for >=2 internal states.

Comment (m1kola) [reply pkg/api/v20191231preview/openshiftcluster.go:89]: Also how do we put the document into `OperatorUpdating` state? Sounds like we need to modify [`putOrPatchOpenShiftCluster`](https://github.com/Azure/ARO-RP/blob/f9eaa7ffe8b6c431325d71e6e7532092ac605d21/pkg/frontend/openshiftcluster_putorpatch.go#L191-L198) somehow (introduce another special case?).

Comment (m1kola) [reply pkg/api/v20191231preview/openshiftcluster.go:89]: What are the alternative ways to implement this?

Maybe we can re-use `AdminUpdating` and parametrise it (e.g. update all, update operator only, etc)?

**Review by m1kola — DISMISSED** [review]
I've got a lot of questions.

File: pkg/api/openshiftcluster.go [hunk 3]
```diff
@@ -151,19 +151,25 @@ type ProvisioningState string
...
-	ProvisioningStateCreating      ProvisioningState = "Creating"
-	ProvisioningStateUpdating      ProvisioningState = "Updating"
-	ProvisioningStateAdminUpdating ProvisioningState = "AdminUpdating"
-	ProvisioningStateDeleting      ProvisioningState = "Deleting"
-	ProvisioningStateSucceeded     ProvisioningState = "Succeeded"
-	ProvisioningStateFailed        ProvisioningState = "Failed"
+	ProvisioningStateCreating         ProvisioningState = "Creating"
+	ProvisioningStateUpdating         ProvisioningState = "Updating"
+	ProvisioningStateAdminUpdating    ProvisioningState = "AdminUpdating"
+	ProvisioningStateOperatorUpdating ProvisioningState = "OperatorUpdating"
```
Comment (mjudeikis) [inline pkg/api/openshiftcluster.go:157]: Why not reuse the state?

File: pkg/api/v20191231preview/openshiftcluster.go (hunk 2 above)
Comment (mjudeikis) [reply pkg/api/v20191231preview/openshiftcluster.go:89]: +1 on re-use the state

**Review by mjudeikis — CHANGES_REQUESTED** [review]
Currently AdminUpdate should not fail as it is failing now.
It should be ephemeral. In example customer-update if ran it read inputs data and execute only actions we need to (we run via all steps but most of them if no input there is noop).
I think we should be doing same for AdminUpdate + add operator tag/hash into admin API and trigger just steps we need. 
Similar as Mikalai suggest but maybe make AdminUpdate more ephemeral, dont fail and emit metrics for failed steps so we can take action different way

File: pkg/api/openshiftcluster.go (hunk 3 above)
Comment (hawkowl) [reply pkg/api/openshiftcluster.go:157]: Because presently we don't parameterise -- although that does sound like a good idea

File: pkg/api/v20191231preview/openshiftcluster.go (hunk 2 above)
Comment (hawkowl) [reply pkg/api/v20191231preview/openshiftcluster.go]: Done on the new version.

File: pkg/api/openshiftcluster.go [hunk 4]
```diff
@@ -160,6 +161,13 @@ const (
 	ProvisioningStateFailed        ProvisioningState = "Failed"
 )
 
+type MaintenanceTask string
+
+const (
+	MaintenanceTaskInfrastructure MaintenanceTask = "Infrastructure"
```
Comment (m1kola) [inline pkg/api/openshiftcluster.go:167]: Something like "everything" instead of `Infrastructure`. As far I see from a quick look - it runs all the stesps.

**Review by m1kola — COMMENTED** [review]
I didn't manage to go through everything. Just first comments.

File: pkg/cluster/cluster.go [hunk 5]
```diff
@@ -88,6 +89,12 @@ type manager struct {
 	samplescli    samplesclient.Interface
 	securitycli   securityclient.Interface
 	arocli        aroclient.Interface
+	dynamichelper dynamichelper.Interface
+
+	// We pass in these clients when we do testing, so add an option to omit
+	// creating them
+	shouldInitAzureClients      bool
```
[the remark this answers is not in the document]
Comment (hawkowl) [reply pkg/cluster/cluster.go:96]: I think we don't and this is a mistake, oops

File: pkg/cluster/install.go [hunk 6]
```diff
@@ -21,41 +21,81 @@ import (
...
+	"github.com/Azure/ARO-RP/pkg/util/dynamichelper"
 	"github.com/Azure/ARO-RP/pkg/util/restconfig"
 	"github.com/Azure/ARO-RP/pkg/util/steps"
 	"github.com/Azure/ARO-RP/pkg/util/version"
 )
 
 // AdminUpdate performs an admin update of an ARO cluster
 func (m *manager) AdminUpdate(ctx context.Context) error {
-	steps := []steps.Step{
+	return m.adminUpdate(ctx, m.runSteps)
```
[the remark this answers is not in the document]
Comment (hawkowl) [reply pkg/cluster/install.go:32]: I think I have a better solution for this. I'll give it a stab.

File: pkg/frontend/openshiftcluster_putorpatch.go [hunk 7]
```diff
@@ -206,6 +206,9 @@ func (f *frontend) _putOrPatchOpenShiftCluster(ctx context.Context, log *logrus.
 		if vars["api-version"] == admin.APIVersion {
 			doc.OpenShiftCluster.Properties.ProvisioningState = api.ProvisioningStateAdminUpdating
 			doc.OpenShiftCluster.Properties.LastAdminUpdateError = ""
+			if doc.OpenShiftCluster.Properties.MaintenanceTask == "" {
+				doc.OpenShiftCluster.Properties.MaintenanceTask = api.MaintenanceTaskInfrastructure
+			}
```
[the remark this answers is not in the document]
Comment (hawkowl) [reply pkg/frontend/openshiftcluster_putorpatch.go:211]: :thinking: Yeah, I think so.

File: pkg/util/steps/condition.go [hunk 8]
```diff
@@ -40,9 +40,11 @@ type conditionStep struct {
 	pollInterval time.Duration
 }
 
-func (c conditionStep) run(ctx context.Context, log *logrus.Entry) error {
+func (c conditionStep) run(ctx context.Context, log *logrus.Entry, timeDilation int) error {
 	var pollInterval time.Duration
-	timeoutCtx, cancel := context.WithTimeout(ctx, c.timeout)
+
+	timeout := time.Duration(c.timeout.Nanoseconds() / int64(timeDilation))
+	timeoutCtx, cancel := context.WithTimeout(ctx, timeout)
```
[the remark this answers is not in the document]
Comment (hawkowl) [reply pkg/util/steps/condition.go:47]: Kind of. It's a "modifier" for how long the timeouts should be, because they're implemented inside the functions. There might be a better option for this.

File: pkg/operator/deploy/deploy.go [hunk 9]
```diff
@@ -65,6 +64,10 @@ func New(log *logrus.Entry, env env.Interface, oc *api.OpenShiftCluster, arocli
 		return nil, err
 	}
 
+	return NewWithProvidedClients(log, env, oc, arocli, extensionscli, kubernetescli, dh), nil
```
[the remark this answers is not in the document]
Comment (hawkowl) [reply pkg/operator/deploy/deploy.go:67]: Good idea.

File: pkg/api/openshiftcluster.go (hunk 4 above)
Comment (hawkowl) [reply pkg/api/openshiftcluster.go:167]: Yeah, that's a good idea.

File: pkg/api/admin/openshiftcluster.go [hunk 10]
```diff
@@ -34,6 +34,7 @@ type OpenShiftClusterProperties struct {
 	LastProvisioningState           ProvisioningState       `json:"lastProvisioningState,omitempty"`
 	FailedProvisioningState         ProvisioningState       `json:"failedProvisioningState,omitempty"`
 	LastAdminUpdateError            string                  `json:"lastAdminUpdateError,omitempty"`
+	MaintenanceTask                 MaintenanceTask         `json:"maintenanceTask,omitempty"`
```
[the remark this answers is not in the document]
Comment (hawkowl) [reply pkg/api/admin/openshiftcluster.go:37]: As in, once we've set it?

File: pkg/cluster/adminupdate_test.go [hunk 11]
```diff
@@ -0,0 +1,331 @@
...
+	mock_env "github.com/Azure/ARO-RP/pkg/util/mocks/env"
+	"github.com/Azure/ARO-RP/pkg/util/steps"
+	utiltls "github.com/Azure/ARO-RP/pkg/util/tls"
+	testdatabase "github.com/Azure/ARO-RP/test/database"
+	testdynamichelper "github.com/Azure/ARO-RP/test/util/dynamichelper"
+	testlog "github.com/Azure/ARO-RP/test/util/log"
+	"github.com/Azure/ARO-RP/test/util/stephooks"
+)
+
+func HEALTHY_APISERVER_OPERATOR() *configv1.ClusterOperator {
```
Comment (mjudeikis) [inline pkg/cluster/adminupdate_test.go:39]: What is this name? lets not add Python-ism into the golang code...

File: pkg/cluster/install.go [hunk 12]
```diff
@@ -21,46 +21,82 @@ import (
...
+
+	// Generic fix-up or setup actions that are fairly safe to always take
+	toRun := []steps.Step{
 		steps.Action(m.initializeKubernetesClients), // must be first
-		steps.Action(m.fixupClusterSPObjectID),
+		steps.Action(m.ensureBillingRecord),         // belt and braces
 		steps.Action(m.ensureDefaults),
-		steps.AuthorizationRefreshingAction(m.fpAuthorizer, steps.Action(m.ensureResourceGroup)), // re-create RP RBAC if needed after tenant migration
-		steps.Action(m.createOrUpdateDenyAssignment),
+		steps.NonCriticalAction(m.fixupClusterSPObjectID),
```
Comment (mjudeikis) [inline pkg/cluster/install.go:40]: `NonCriticalAction`  change does not look related to this PR. 
Why are we doing this in the first place?
Like DenyAssignment is for sure critical. This is very confusing now. Very! :D

File: pkg/cluster/install.go [hunk 13]
```diff
@@ -21,46 +21,82 @@ import (
...
+		)
+	}
+
+	return m.runSteps(ctx, toRun)
 }
 
 func (m *manager) Update(ctx context.Context) error {
 	steps := []steps.Step{
-		steps.AuthorizationRefreshingAction(m.fpAuthorizer, steps.Action(m.validateResources)),
+		steps.AuthorizationRefreshingAction(m.fpAuthorizer, m.applyTimeDilation(10*time.Minute), steps.Action(m.validateResources)),
```
Comment (mjudeikis) [inline pkg/cluster/install.go:99]: Why do we add timeDilatation? Again, does not look like part of this PR but might be wrong

**Review by mjudeikis — COMMENTED** [review]
This is very big and quite complex PR.
Can we split it? like `kruntime` changes are irrelevant and makes majority if this PR. Separate PR where we can just merge would be good. 
There is a lot of other things like applyTimeDilation and NonCriticalAction where reason for it is not quite clear. Im ok to leave in this PR IF this needs to go for it to be working, but this does not look like a case :/

File: pkg/cluster/adminupdate_test.go (hunk 11 above)
Comment (hawkowl) [reply pkg/cluster/adminupdate_test.go]: ;D fixed

Old habits die hard...

File: pkg/cluster/install.go (hunk 13 above)
Comment (hawkowl) [reply pkg/cluster/install.go]: Re-thought how to do this, fixed

**Comment by hawkowl** [comment]
I've split out things that I don't see as directly needed to write the unit tests related to the code.

File: pkg/api/admin/openshiftcluster.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (m1kola) [reply pkg/api/admin/openshiftcluster.go:37]: This might prevent us from changing this field via admin API:

https://github.com/Azure/ARO-RP/blob/fef8f41aa37faa8bae0e27ea896a40a549cd9d3a/pkg/api/admin/openshiftcluster_validatestatic.go#L26-L30

I can imagine a case of update failing with `maintenanceTask` set to `OperatorUpdate`. How do we fix it? Will we have to manually edit a doc in the DB and remove this field?

[the remark this answers is not in the document]
Comment (hawkowl) [reply pkg/api/admin/openshiftcluster.go:37]: OHHHHHHH

I get you now. I thought you were asking if it had to be mutable, in "should make it immutable", I forgot that it was immutable by default.

Fixed.

**Comment by hawkowl** [comment]
@m1kola I made sure that it gets cleared when EndLease happens. Then, if it gets stuck (because the lease expired?) we can just do another admin update to re-clear it.

I also investigated a bit further, and it turned out that the Admin API tests weren't using the static validator, and so my new test as well as another new one were passing falsely. I fixed up that test and now I have validated that we can change the field.

File: pkg/frontend/openshiftcluster_putorpatch_test.go [hunk 14]
```diff
@@ -141,8 +300,6 @@ func TestPutOrPatchOpenShiftClusterAdminAPI(t *testing.T) {
 		{
 			name: "patch a cluster with registry profile should ignore registry profile",
 			request: func(oc *admin.OpenShiftCluster) {
-				oc.Properties.ClusterProfile.Domain = "changed"
-				oc.Name = "resourceName"
```
Comment (m1kola) [inline pkg/frontend/openshiftcluster_putorpatch_test.go:145]: I don't follow why we change this

File: pkg/cluster/arooperator.go [hunk 15]
```diff
@@ -10,22 +10,13 @@ import (
 )
 
 func (m *manager) ensureAROOperator(ctx context.Context) error {
-	dep, err := deploy.New(m.log, m.env, m.doc.OpenShiftCluster, m.arocli, m.extensionscli, m.kubernetescli)
-	if err != nil {
-		m.log.Errorf("cannot ensureAROOperator.New: %s", err.Error())
-		return err
-	}
-	err = dep.CreateOrUpdate(ctx)
+	err := deploy.New(m.log, m.env, m.doc.OpenShiftCluster, m.arocli, m.extensionscli, m.kubernetescli, m.dynamichelper).CreateOrUpdate(ctx)
```
Comment (m1kola) [inline pkg/cluster/arooperator.go:13]: Please let's not use chaining in Go code. Especially in such a long line. It is hard to follow: took me a minute to understand why we call `deploy.New()`, but the error is about `CreateOrUpdate`.

File: pkg/api/admin/openshiftcluster.go [hunk 16]
```diff
@@ -48,7 +49,7 @@ type OpenShiftClusterProperties struct {
 	IngressProfiles                 []IngressProfile        `json:"ingressProfiles,omitempty"`
 	Install                         *Install                `json:"install,omitempty"`
 	StorageSuffix                   string                  `json:"storageSuffix,omitempty"`
-	RegistryProfiles                []RegistryProfile       `json:"registryProfiles,omitempty"`
+	RegistryProfiles                []RegistryProfile       `json:"registryProfiles,omitempty" mutable:"true"` // Not actually mutable! We discard it in ToInternal, this just means that any changes do not cause static validation to fail.
```
Comment (m1kola) [inline pkg/api/admin/openshiftcluster.go:52]: I don't understand this change (even with the comment). Are you saying that there was a bug in the existing code? If so - could you please submit a separate PR fixing only this issue? Side effects of this change are not clear to me from the `+1,576 −172` diff.

File: pkg/frontend/openshiftcluster_putorpatch_test.go (hunk 14 above)
Comment (hawkowl) [reply pkg/frontend/openshiftcluster_putorpatch_test.go:145]: See https://github.com/Azure/ARO-RP/pull/1803

File: pkg/cluster/arooperator.go (hunk 15 above)
Comment (hawkowl) [reply pkg/cluster/arooperator.go]: done :)

File: pkg/cluster/cluster.go [hunk 17]
```diff
@@ -90,6 +91,11 @@ type manager struct {
 	securitycli    securityclient.Interface
 	arocli         aroclient.Interface
 	registryclient imageregistryclient.Interface
+	dynamichelper  dynamichelper.Interface
```
Comment (m1kola) [inline pkg/cluster/cluster.go:94]: Where do we instantiate it? I do not see it in `New`.

File: pkg/database/openshiftclusters.go [hunk 18]
```diff
@@ -306,6 +306,7 @@ func (c *openShiftClusters) EndLease(ctx context.Context, key string, provisioni
 	return c.patchWithLease(ctx, key, func(doc *api.OpenShiftClusterDocument) error {
 		doc.OpenShiftCluster.Properties.ProvisioningState = provisioningState
 		doc.OpenShiftCluster.Properties.FailedProvisioningState = failedProvisioningState
+		doc.OpenShiftCluster.Properties.MaintenanceTask = ""
```
Comment (m1kola) [inline pkg/database/openshiftclusters.go:309]: Wondering if `endLease` in `pkg/backend/openshiftcluster.go` is going to be a more appropriate place for this logic?

File: pkg/frontend/openshiftcluster_putorpatch.go [hunk 19]
```diff
@@ -202,6 +199,9 @@ func (f *frontend) _putOrPatchOpenShiftCluster(ctx context.Context, log *logrus.
 		doc.Dequeues = 0
 	}
 
+	// SetDefaults will set defaults on cluster document
+	api.SetDefaults(doc)
```
Comment (m1kola) [inline pkg/frontend/openshiftcluster_putorpatch.go:203]: Why move?

File: pkg/cluster/cluster.go [hunk 20]
```diff
@@ -90,6 +91,11 @@ type manager struct {
...
+
+	// We pass in these clients when we do testing, so add an option to omit
+	// creating them
+	shouldInitKubernetesClients bool
```
Comment (m1kola) [inline pkg/cluster/cluster.go:98]: Something is wrong if we have to do it, but I haven't figured out what exactly yet. Still a lot to review.

File: pkg/util/steps/runner.go [hunk 21]
```diff
@@ -20,18 +21,42 @@ func friendlyName(f interface{}) string {
...
 	run(ctx context.Context, log *logrus.Entry) error
+	setPollInterval(time.Duration)
+	setTimeout(time.Duration)
 	String() string
 }
 
 // Run executes the provided steps in order until one fails or all steps
 // are completed. Errors from failed steps are returned directly.
-func Run(ctx context.Context, log *logrus.Entry, pollInterval time.Duration, steps []Step) error {
+func Run(ctx context.Context, log *logrus.Entry, pollInterval time.Duration, globalTimeout time.Duration, steps []Step, testHook TestHook) error {
```
Comment (m1kola) [inline pkg/util/steps/runner.go:31]: I think we are doing someting wrong if in the production code we have to add an extra param for testing porpuses. I still have a lot to review and I need to dig more into this. Hopefully I'll be able to suggest some alternative way.

File: pkg/cluster/install.go [hunk 22]
```diff
@@ -29,35 +29,75 @@ import (
 
 // AdminUpdate performs an admin update of an ARO cluster
 func (m *manager) AdminUpdate(ctx context.Context) error {
-	steps := []steps.Step{
+	toRun := m.adminUpdate(ctx)
+	return m.runSteps(ctx, toRun)
+}
+
+func (m *manager) adminUpdate(ctx context.Context) []steps.Step {
```
Comment (bennerv) [inline pkg/cluster/install.go:36]: ctx is not used at all in this function
```suggestion
func (m *manager) adminUpdate() []steps.Step {
```

File: pkg/cluster/install.go [hunk 23]
```diff
@@ -29,35 +29,75 @@ import (
...
+
+	// We don't run this on an operator-only deploy as PUCM scripts then cannot
+	// determine if the cluster has been fully admin-updated
+	if isEverything {
+		toRun = append(toRun,
+			steps.Action(m.updateProvisionedBy), // Run this last so we capture the resource provider only once the upgrade has been fully performed
+		)
+	}
+
+	return toRun
```
Comment (bennerv) [inline pkg/cluster/install.go:100]: I'm wondering if there's a better way to clean this up a bit.  

Even with the comments, it can be confusing for someone working in this spot to understand that every step is in a specific order, and part of a whole (either everything, or operator maintenance) and when to use what.  

Just general comment.

**Review by bennerv — CHANGES_REQUESTED** [review]
One minor nit, another comment in general.

otherwise makes sense and LGTM.

File: pkg/cluster/install.go (hunk 22 above)
Comment (hawkowl) [reply pkg/cluster/install.go]: fixed

**Review by bennerv — APPROVED** [review]
LGTM
